torch/lib/c10d/Utils.hpp

```diff
@@ -292,8 +292,10 @@ using SizeType = uint64_t;
 #define SYSCHECK(expr)                                        \
   {                                                           \
     errno = 0;                                                \
-    auto ___output = (expr);                                  \
-    (void)___output;                                          \
+    while (errno != EINTR) {                                  \
```

Comment:
an unlimited while loop is dangerous. Switch to having a max number of retries

Reply:
This is probably not the correct strategy since dist tests are failing. Probably what I should do is to not use `SYSCHECK` and do a manual retry at https://github.com/pytorch/pytorch/blob/2681852438f9f80b43fd2502d83839140733b68c/torch/lib/c10d/TCPStore.cpp#L70. Or maybe a non-throwing version of `SYSCHECK`? Need to hear from the code owners @teng-li and @pietern .

Reply:
Unlimited retry on EINTR is the way to go... A syscall may be interrupted many times especially if it is long running (e.g. think of `poll` without timeout to signal some rare event). I found a Python PEP that changed Python to do this at the lowest syscall wrapper level, see https://www.python.org/dev/peps/pep-0475/#implementation. The difference between Python and here is that in Python you need to check for Python-side exceptions in the signal handlers before retrying the system call. We don't have an equivalent to that and rely on Python shutting down to tear down all C++ side.